train_network.py

```diff
@@ -0,0 +1,179 @@
+next_element = iterator.get_next()
+training_iterator = train_data.make_initializable_iterator()
+val_iterator = test_data.make_initializable_iterator()
+
+# create placeholder for train or test
+train_network = tf.placeholder(tf.bool, [])
+
+# get images and pass into network
+image, label, weight = next_element
+drn = DRN(image, image_dims, batch_size, num_classes, train_network, network)
+
+# get predictions and logits
+prediction = drn.pred
+logits = drn.prob
+label = tf.squeeze(label, 3)
+
+# resize the logits using bilinear interpolation
+imsize = tf.constant([iamge_dims[0], image_dims[1]], dtype=tf.int32)
+logits = tf.image.resize_bilinear(logits, imsize)
+prediction = tf.argmax(logits, 3)
+print('Resized shape is {}'.format(logits.get_shape()))
+
+ 
+# compute loss
+if use_weights == 1:
+  label_one_hot = tf.one_hot(label, num_classes)
+  loss = tf.nn.softmax_cross_entropy_with_logits(labels=label_one_hot, logits=logits)
+  loss = loss*tf.squeeze(weight, 3)
+else:
+  # use sparse with flattened labelmaps
+  loss = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=label, logits=logits)
+loss = tf.reduce_mean(loss)
+
+#loss = dice(logits, label, num_classes, use_weights=1)
+
+# create summary
+tf.summary.scalar('loss', loss)
+tf.summary.image('images', image)
+tf.summary.image('predictions', tf.cast(tf.expand_dims(prediction, 3), dtype=tf.uint8))
+tf.summary.image('labels', tf.cast(tf.expand_dims(label, 3), dtype=tf.uint8))
+
+
+# add weights with of first layer
+print(tf.trainable_variables())
+for var in tf.trainable_variables():
+  if'_1' in var.name:
+    tf.summary.histogram(var.name, var)
+#tf.summary.scalar('learning_rate', learning_rate)
+merged = tf.summary.merge_all()
+train_writer = tf.summary.FileWriter(logs, tf.get_default_graph())
+
+# global step to keep track of iterations
+global_step = tf.Variable(0, trainable=False, name='global_step')
+
+# create placeholder for learning rate
+learning_rate = tf.placeholder(tf.float32, shape=[])
+
+# training 
+training = tf.train.AdamOptimizer(learning_rate).minimize(loss, global_step)
+
+saver = tf.train.Saver(max_to_keep=3)
+init = tf.global_variables_initializer()
+
+plt.figure(figsize=(18, 16))
+config = tf.ConfigProto()
+config.gpu_options.allow_growth = True
+
+with tf.Session(config=config) as sess:
+  training_handle = sess.run(training_iterator.string_handle())
+  validation_handle = sess.run(val_iterator.string_handle())
+  
+  sess.run(training_iterator.initializer)
+  # initialize variables 
+  sess.run(init)
+  
+  # check if checkpiont exists
+  ckpt = tf.train.get_checkpoint_state(logs)
+  if ckpt and ckpt.model_checkpoint_path:
+    saver.restore(sess, ckpt.model_checkpoint_path) 
+    print('Restoring session at step {}'.format(global_step.eval()))
+    
+  iteration = global_step.eval()
+  current_epoch = int(np.ceil(iteration/training_iterations)) 
+  
+  while current_epoch < num_epochs:
```

Comment:
`while` loops are used when you don't know how long the loop should continue for, otherwise `for` loops are appropriate. Also done again on L132. `for i in range(training_iterations)` is a pretty common usage (also done this way on L155, which is odd it's not done here)